pkg/nimble/netif/nimble_netif.c

```diff
@@ -0,0 +1,630 @@
+        (GNRC_NETIF_HDR_FLAGS_BROADCAST | GNRC_NETIF_HDR_FLAGS_MULTICAST)) {
+        nimble_netif_conn_foreach(NIMBLE_NETIF_L2CAP_CONNECTED,
+                                  _netif_send_iter, pkt->next);
+        res = (int)gnrc_pkt_len(pkt->next);
+    }
+    /* send unicast */
+    else {
+        int handle = nimble_netif_conn_get_by_addr(
+            gnrc_netif_hdr_get_dst_addr(hdr));
+        nimble_netif_conn_t *conn = nimble_netif_conn_get(handle);
+        res = _send_pkt(conn, pkt->next);
+    }
+
+    /* release the packet in GNRC's packet buffer */
+    gnrc_pktbuf_release(pkt);
+    return res;
+}
+
+/* not used, we pass incoming data to GNRC directly from the NimBLE thread */
+static gnrc_pktsnip_t *_netif_recv(gnrc_netif_t *netif)
+{
+    (void)netif;
+    return NULL;
+}
+
+static const gnrc_netif_ops_t _nimble_netif_ops = {
+    .init = _netif_init,
+    .send = _netif_send,
+    .recv = _netif_recv,
+    .get = gnrc_netif_get_from_netdev,
+    .set = gnrc_netif_set_from_netdev,
+    .msg_handler = NULL,
+};
+
+static inline int _netdev_init(netdev_t *dev)
+{
+    _nimble_netif = dev->context;
+
+    /* get our own address from the controller */
+    uint8_t tmp[6];
+    int res = ble_hs_id_copy_addr(nimble_riot_own_addr_type, tmp, NULL);
+    assert(res == 0);
+    (void)res;
+
+    bluetil_addr_swapped_cp(tmp, _nimble_netif->l2addr);
+    return 0;
+}
+
+static inline int _netdev_get(netdev_t *dev, netopt_t opt,
+                              void *value, size_t max_len)
+{
+    (void)dev;
+    int res = -ENOTSUP;
+
+    switch (opt) {
+        case NETOPT_ADDRESS:
+            assert(max_len >= BLE_ADDR_LEN);
+            memcpy(value, _nimble_netif->l2addr, BLE_ADDR_LEN);
+            res = BLE_ADDR_LEN;
+            break;
+        case NETOPT_ADDR_LEN:
+        case NETOPT_SRC_LEN:
+            assert(max_len == sizeof(uint16_t));
+            *((uint16_t *)value) = BLE_ADDR_LEN;
+            res = sizeof(uint16_t);
+            break;
+        case NETOPT_MAX_PACKET_SIZE:
+            assert(max_len >= sizeof(uint16_t));
+            *((uint16_t *)value) = MTU_SIZE;
+            res = sizeof(uint16_t);
+            break;
+        case NETOPT_PROTO:
+            assert(max_len == sizeof(gnrc_nettype_t));
+            *((gnrc_nettype_t *)value) = _nettype;
+            res = sizeof(gnrc_nettype_t);
+            break;
+        case NETOPT_DEVICE_TYPE:
+            assert(max_len == sizeof(uint16_t));
+            *((uint16_t *)value) = NETDEV_TYPE_BLE;
+            res = sizeof(uint16_t);
+            break;
+        default:
+            break;
+    }
+
+    return res;
+}
+
+static inline int _netdev_set(netdev_t *dev, netopt_t opt,
+                              const void *value, size_t val_len)
+{
+    (void)dev;
+    int res = -ENOTSUP;
+
+    switch (opt) {
+        case NETOPT_PROTO:
+            assert(val_len == sizeof(gnrc_nettype_t));
+            memcpy(&_nettype, value, sizeof(gnrc_nettype_t));
```

Comment:
Maybe premature optimization, but I think an assignment would safe some ROM.

```suggestion
            _nettype = *((gnrc_nettype_t *)value);
```

Otherwise, I would prefer to use `sizeof(_nettype)` instead for the size argument.

Reply:
ROM wise, it does not make any difference. But did adapt to use `sizeof(_nettype)`.

Reply:
Ok.